cnf-certification-test/lifecycle/scaling/statefulset_scaling.go

```diff
@@ -0,0 +1,194 @@
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+//nolint:dupl
+package scaling
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/test-network-function/cnf-certification-test/cnf-certification-test/lifecycle/podsets"
+	"github.com/test-network-function/cnf-certification-test/internal/clientsholder"
+
+	v1app "k8s.io/api/apps/v1"
+	v1autoscaling "k8s.io/api/autoscaling/v1"
+	v1 "k8s.io/client-go/kubernetes/typed/apps/v1"
+
+	v1machinery "k8s.io/apimachinery/pkg/apis/meta/v1"
+	retry "k8s.io/client-go/util/retry"
+
+	hps "k8s.io/client-go/kubernetes/typed/autoscaling/v1"
+)
+
+func TestScaleStateFulset(statefulset *v1app.StatefulSet, timeout time.Duration) bool {
+	clients := clientsholder.GetClientsHolder()
+	name, namespace := statefulset.Name, statefulset.Namespace
+	ssClients := clients.AppsClients.StatefulSets(namespace)
+	logrus.Trace("scale statefulset not using HPA ", namespace, ":", name)
+	var replicas int32
+	if statefulset.Spec.Replicas != nil {
+		replicas = *statefulset.Spec.Replicas
+	} else {
+		replicas = 1
+	}
+
+	if replicas <= 1 {
+		// scale up
+		replicas++
+		if !scaleStateFulsetHelper(clients, ssClients, statefulset, replicas, timeout, true) {
+			logrus.Error("can't scale statefulset =", namespace, ":", name)
+			return false
+		}
+		// scale down
+		replicas--
+		if !scaleStateFulsetHelper(clients, ssClients, statefulset, replicas, timeout, false) {
+			logrus.Error("can't scale statefulset =", namespace, ":", name)
+			return false
+		}
+	} else {
+		// scale down
+		replicas--
+		if !scaleStateFulsetHelper(clients, ssClients, statefulset, replicas, timeout, false) {
+			logrus.Error("can't scale statefulset =", namespace, ":", name)
+			return false
+		} // scale up
+		replicas++
+		if !scaleStateFulsetHelper(clients, ssClients, statefulset, replicas, timeout, true) {
+			logrus.Error("can't scale statefulset =", namespace, ":", name)
+			return false
+		}
+	}
+	return true
+}
+
+func scaleStateFulsetHelper(clients *clientsholder.ClientsHolder, ssClient v1.StatefulSetInterface, statefulset *v1app.StatefulSet, replicas int32, timeout time.Duration, up bool) bool {
+	if up {
+		logrus.Trace("scale UP statefulset to ", replicas, " replicas ")
+	} else {
+		logrus.Trace("scale DOWN statefulset to ", replicas, " replicas ")
+	}
+
+	name := statefulset.Name
+	namespace := statefulset.Namespace
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		// Retrieve the latest version of statefulset before attempting update
+		// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+		ss, err := ssClient.Get(context.TODO(), name, v1machinery.GetOptions{})
+		if err != nil {
+			logrus.Error("failed to get latest version of statefulset ", namespace, ":", name)
```

Comment:
Please use tnf.ClaimFilePrintf for this errors, please. It will be easier to debug issues, as sometimes the tnf log is not accessible. Also, include the err value in the trace.